diff -Naur -Naur mysql-4.1.18.orig/libmysql/Makefile.shared mysql-4.1.18.new/libmysql/Makefile.shared
--- mysql-4.1.18.orig/libmysql/Makefile.shared	2006-01-27 16:26:16.000000000 +0000
+++ mysql-4.1.18.new/libmysql/Makefile.shared	2006-04-10 17:36:35.000000000 +0000
@@ -81,7 +81,8 @@
 			$(target)
 DEFS =			-DDEFAULT_CHARSET_HOME="\"$(MYSQLBASEdir)\"" \
 			-DDATADIR="\"$(MYSQLDATAdir)\"" \
-			-DSHAREDIR="\"$(MYSQLSHAREdir)\"" $(target_defs)
+			-DSHAREDIR="\"$(MYSQLSHAREdir)\"" \
+			-DDEFAULT_SYSCONFDIR="\"$(sysconfdir)\"" $(target_defs)
 
 # The automatic dependencies miss this
 #bmove_upp.lo:		$(LTCHARSET_OBJS)
diff -Naur -Naur mysql-4.1.18.orig/mysys/default.c mysql-4.1.18.new/mysys/default.c
--- mysql-4.1.18.orig/mysys/default.c	2006-01-27 16:26:04.000000000 +0000
+++ mysql-4.1.18.new/mysys/default.c	2006-04-10 18:51:43.000000000 +0000
@@ -43,7 +43,9 @@
 /* Which directories are searched for options (and in which order) */
 
 const char *default_directories[]= {
-#ifdef __WIN__
+#ifdef DEFAULT_SYSCONFDIR
+DEFAULT_SYSCONFDIR,
+#elif defined(__WIN__)
 "C:/",
 #elif defined(__NETWARE__)
 "sys:/etc/",
diff -Naur -Naur mysql-4.1.18.orig/mysys/Makefile.am mysql-4.1.18.new/mysys/Makefile.am
--- mysql-4.1.18.orig/mysys/Makefile.am	2006-01-27 16:26:26.000000000 +0000
+++ mysql-4.1.18.new/mysys/Makefile.am	2006-04-10 17:35:42.000000000 +0000
@@ -68,6 +68,7 @@
 			-DDATADIR="\"$(MYSQLDATAdir)\"" \
 			-DDEFAULT_CHARSET_HOME="\"$(MYSQLBASEdir)\"" \
 			-DSHAREDIR="\"$(MYSQLSHAREdir)\"" \
+			-DDEFAULT_SYSCONFDIR="\"$(sysconfdir)\"" \
                         @DEFS@
 
 libmysys_a_DEPENDENCIES= @THREAD_LOBJECTS@
