diff -Naur mysql-4.1.15/mysys/Makefile.am mysql-4.1.15.new/mysys/Makefile.am
--- mysql-4.1.15/mysys/Makefile.am	2005-10-09 22:46:29.000000000 +0000
+++ mysql-4.1.15.new/mysys/Makefile.am	2005-11-21 13:48:31.000000000 +0000
@@ -68,6 +68,7 @@
 			-DDATADIR="\"$(MYSQLDATAdir)\"" \
 			-DDEFAULT_CHARSET_HOME="\"$(MYSQLBASEdir)\"" \
 			-DSHAREDIR="\"$(MYSQLSHAREdir)\"" \
+			-DDEFAULT_SYSCONFDIR="\"$(sysconfdir)\"" \
                         @DEFS@
 
 libmysys_a_DEPENDENCIES= @THREAD_LOBJECTS@
diff -Naur mysql-4.1.15/mysys/default.c mysql-4.1.15.new/mysys/default.c
--- mysql-4.1.15/mysys/default.c	2005-10-09 22:46:25.000000000 +0000
+++ mysql-4.1.15.new/mysys/default.c	2005-11-21 13:47:11.000000000 +0000
@@ -43,6 +43,9 @@
 /* Which directories are searched for options (and in which order) */
 
 const char *default_directories[]= {
+#ifdef DEFAULT_SYSCONFDIR
+DEFAULT_SYSCONFDIR,
+#endif
 #ifdef __WIN__
 "C:/",
 #elif defined(__NETWARE__)
