diff -urN mysql-orig/configure.in mysql/configure.in
--- mysql-orig/configure.in	2010-01-16 15:17:39.991476010 -0100
+++ mysql/configure.in	2010-01-16 15:17:50.272599930 -0100
@@ -227,7 +227,7 @@
 fi
 
 # Still need ranlib for readline; local static use only so no libtool.
-AC_PROG_RANLIB
+#AC_PROG_RANLIB
 # We use libtool
 #AC_LIBTOOL_WIN32_DLL
 AC_PROG_LIBTOOL
@@ -2779,13 +2779,11 @@
 AC_SUBST(netware_dir)
 AM_CONDITIONAL(HAVE_NETWARE, test "$netware_dir" = "netware")
 
-if test "$with_server" = "yes" -o "$THREAD_SAFE_CLIENT" != "no"
+AM_CONDITIONAL([NEED_THREADS], [test x"$with_server" != xno -o x"$THREAD_SAFE_CLIENT" != xno])
+if test "$with_server" != "no" -o "$THREAD_SAFE_CLIENT" != "no"
 then
   AC_DEFINE([THREAD], [1],
             [Define if you want to have threaded code. This may be undef on client code])
-  # Avoid _PROGRAMS names
-  THREAD_LOBJECTS="thr_alarm.o thr_lock.o thr_mutex.o thr_rwlock.o my_pthread.o my_thr_init.o mf_keycache.o"
-  AC_SUBST(THREAD_LOBJECTS)
   server_scripts="mysqld_safe mysql_install_db"
   sql_server_dirs="strings mysys dbug extra regex"
 
diff -urN mysql-orig/dbug/Makefile.am mysql/dbug/Makefile.am
--- mysql-orig/dbug/Makefile.am	2010-01-16 15:17:39.958474837 -0100
+++ mysql/dbug/Makefile.am	2010-01-16 15:17:50.272599930 -0100
@@ -16,10 +16,10 @@
 # MA 02111-1307, USA
 
 INCLUDES =              -I$(top_builddir)/include -I$(top_srcdir)/include
-LDADD =                 libdbug.a ../mysys/libmysys.a ../strings/libmystrings.a
-pkglib_LIBRARIES =      libdbug.a
+LDADD =                 libdbug.la ../mysys/libmysys.la ../strings/libmystrings.la
+pkglib_LTLIBRARIES =      libdbug.la
 noinst_HEADERS =        dbug_long.h
-libdbug_a_SOURCES =     dbug.c sanity.c
+libdbug_la_SOURCES =     dbug.c sanity.c
 EXTRA_DIST =            CMakeLists.txt example1.c example2.c example3.c \
                         user.r monty.doc dbug_add_tags.pl \
                         my_main.c main.c factorial.c dbug_analyze.c \
diff -urN mysql-orig/extra/Makefile.am mysql/extra/Makefile.am
--- mysql-orig/extra/Makefile.am	2010-01-16 15:17:39.958474837 -0100
+++ mysql/extra/Makefile.am	2010-01-16 15:17:50.272599930 -0100
@@ -15,8 +15,8 @@
 
 INCLUDES =		-I$(top_builddir)/include -I$(top_srcdir)/include \
 			-I$(top_srcdir)/sql
-LDADD =			@CLIENT_EXTRA_LDFLAGS@ ../mysys/libmysys.a \
-			../dbug/libdbug.a ../strings/libmystrings.a \
+LDADD =			@CLIENT_EXTRA_LDFLAGS@ ../mysys/libmysys.la \
+			../dbug/libdbug.la ../strings/libmystrings.la \
 			$(ZLIB_LIBS)
 BUILT_SOURCES=		$(top_builddir)/include/mysqld_error.h \
                         $(top_builddir)/include/sql_state.h \
diff -urN mysql-orig/libmysqld/Makefile.am mysql/libmysqld/Makefile.am
--- mysql-orig/libmysqld/Makefile.am	2010-01-16 15:17:39.868474308 -0100
+++ mysql/libmysqld/Makefile.am	2010-01-16 15:17:50.274371260 -0100
@@ -36,8 +36,8 @@
 			$(openssl_includes) @ZLIB_INCLUDES@ \
 			@condition_dependent_plugin_includes@
 
-noinst_LIBRARIES =	libmysqld_int.a
-pkglib_LIBRARIES =	libmysqld.a
+noinst_LTLIBRARIES =	libmysqld_int.la
+pkglib_LTLIBRARIES =	libmysqld.la
 SUBDIRS =		. examples
 libmysqld_sources=	libmysqld.c lib_sql.cc emb_qcache.cc
 libmysqlsources =	errmsg.c get_password.c libmysql.c client.c pack.c \
@@ -78,9 +78,9 @@
 	rpl_injector.cc my_user.c partition_info.cc \
 	sql_servers.cc event_parse_data.cc
 
-libmysqld_int_a_SOURCES= $(libmysqld_sources)
-nodist_libmysqld_int_a_SOURCES= $(libmysqlsources) $(sqlsources)
-libmysqld_a_SOURCES=
+libmysqld_int_la_SOURCES= $(libmysqld_sources)
+nodist_libmysqld_int_la_SOURCES= $(libmysqlsources) $(sqlsources)
+libmysqld_la_SOURCES=
 
 sqlstoragesources =	$(EXTRA_libmysqld_a_SOURCES)
 storagesources = @condition_dependent_plugin_modules@
@@ -90,19 +90,19 @@
 # automake misses these
 sql_yacc.cc sql_yacc.h: $(top_srcdir)/sql/sql_yacc.yy
 
-# The following libraries should be included in libmysqld.a
-INC_LIB=	$(top_builddir)/regex/libregex.a \
-		$(top_builddir)/mysys/libmysys.a \
-		$(top_builddir)/strings/libmystrings.a \
-		$(top_builddir)/dbug/libdbug.a \
-		$(top_builddir)/vio/libvio.a \
+# The following libraries should be included in libmysqld.la
+INC_LIB=	$(top_builddir)/regex/libregex.la \
+		$(top_builddir)/mysys/libmysys.la \
+		$(top_builddir)/strings/libmystrings.la \
+		$(top_builddir)/dbug/libdbug.la \
+		$(top_builddir)/vio/libvio.la \
                 @NDB_SCI_LIBS@ \
 		@mysql_plugin_libs@ \
 		$(yassl_inc_libs)
 
 if HAVE_YASSL
-yassl_inc_libs=	$(top_builddir)/extra/yassl/src/.libs/libyassl.a \
-		$(top_builddir)/extra/yassl/taocrypt/src/.libs/libtaocrypt.a
+yassl_inc_libs=	$(top_builddir)/extra/yassl/src/.libs/libyassl.la \
+		$(top_builddir)/extra/yassl/taocrypt/src/.libs/libtaocrypt.la
 endif
 
 # Storage engine specific compilation options
@@ -140,7 +140,7 @@
 
 #
 # To make it easy for the end user to use the embedded library we
-# generate a total libmysqld.a from all library files,
+# generate a total libmysqld.la from all library files,
 
 # note - InnoDB libraries have circular dependencies, so in INC_LIB
 # few libraries are present two times. Metrowerks linker doesn't like
@@ -148,7 +148,7 @@
 # need to add the same file twice to the library, so 'sort -u' save us
 # some time and spares unnecessary work.
 
-libmysqld.a:	libmysqld_int.a $(INC_LIB) $(libmysqld_a_DEPENDENCIES) $(storageobjects)
+libmysqld.la:	libmysqld_int.la $(INC_LIB) $(libmysqld_a_DEPENDENCIES) $(storageobjects)
 if DARWIN_MWCC
 	mwld -lib -o $@ libmysqld_int.a `echo $(INC_LIB) | sort -u` $(storageobjects)
 else
diff -urN mysql-orig/mysys/Makefile.am mysql/mysys/Makefile.am
--- mysql-orig/mysys/Makefile.am	2010-01-16 15:17:39.961474618 -0100
+++ mysql/mysys/Makefile.am	2010-01-16 15:23:18.775352467 -0100
@@ -18,10 +18,9 @@
 MYSQLBASEdir=		$(prefix)
 INCLUDES =		@ZLIB_INCLUDES@ -I$(top_builddir)/include \
 			-I$(top_srcdir)/include -I$(srcdir)
-pkglib_LIBRARIES =	libmysys.a
-LDADD =			libmysys.a $(top_builddir)/strings/libmystrings.a $(top_builddir)/dbug/libdbug.a
+pkglib_LTLIBRARIES =	libmysys.la
 noinst_HEADERS =	mysys_priv.h my_static.h my_handler_errors.h
-libmysys_a_SOURCES =    my_init.c my_getwd.c mf_getdate.c my_mmap.c \
+libmysys_la_SOURCES =    my_init.c my_getwd.c mf_getdate.c my_mmap.c \
 			mf_path.c mf_loadpath.c my_file.c \
 			my_open.c my_create.c my_dup.c my_seek.c my_read.c \
 			my_pread.c my_write.c my_getpagesize.c \
@@ -46,18 +45,22 @@
 			my_quick.c my_lockmem.c my_static.c \
 			my_sync.c my_getopt.c my_mkdir.c \
 			default_modify.c default.c \
-                        my_compress.c checksum.c \
+			my_compress.c checksum.c \
 			my_net.c my_port.c my_sleep.c \
 			charset.c charset-def.c my_bitmap.c my_bit.c md5.c \
 			my_gethostbyname.c rijndael.c my_aes.c sha1.c \
 			my_handler.c my_netware.c my_largepage.c \
 			my_memmem.c stacktrace.c \
 			my_windac.c my_access.c base64.c my_libwrap.c
+libmysys_la_LIBADD = $(top_builddir)/dbug/libdbug.la $(top_builddir)/strings/libmystrings.la \
+			$(ZLIB_LIBS)
+if NEED_THREADS
+	libmysys_la_SOURCES += thr_alarm.c thr_lock.c thr_mutex.c thr_rwlock.c my_pthread.c my_thr_init.c
+endif
 EXTRA_DIST =		thr_alarm.c thr_lock.c my_pthread.c my_thr_init.c \
 			thr_mutex.c thr_rwlock.c \
 			CMakeLists.txt mf_soundex.c \
 			my_conio.c my_wincond.c my_winthread.c
-libmysys_a_LIBADD =	@THREAD_LOBJECTS@
 # test_dir_DEPENDENCIES=	$(LIBRARIES)
 # testhash_DEPENDENCIES=	$(LIBRARIES)
 # test_charset_DEPENDENCIES=	$(LIBRARIES)
@@ -71,8 +74,6 @@
 			-DDEFAULT_SYSCONFDIR="\"$(sysconfdir)\"" \
                         @DEFS@
 
-libmysys_a_DEPENDENCIES= @THREAD_LOBJECTS@
-
 # I hope this always does the right thing. Otherwise this is only test programs
 FLAGS=$(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS) @NOINST_LDFLAGS@
 
@@ -81,47 +82,47 @@
 # which automaticly removes the object files you use to compile a final program
 #
 
-test_bitmap$(EXEEXT): my_bitmap.c $(LIBRARIES)
+test_bitmap$(EXEEXT): my_bitmap.c $(LTLIBRARIES)
 	$(LINK) $(FLAGS) -DMAIN  ./my_bitmap.c $(LDADD) $(LIBS)
 
-test_priority_queue$(EXEEXT): queues.c $(LIBRARIES)
+test_priority_queue$(EXEEXT): queues.c $(LTLIBRARIES)
 	$(LINK) $(FLAGS) -DMAIN  ./queues.c $(LDADD) $(LIBS)
 
-test_thr_alarm$(EXEEXT): thr_alarm.c $(LIBRARIES)
+test_thr_alarm$(EXEEXT): thr_alarm.c $(LTLIBRARIES)
 	$(CP) $(srcdir)/thr_alarm.c ./test_thr_alarm.c
 	$(LINK) $(FLAGS) -DMAIN  ./test_thr_alarm.c $(LDADD) $(LIBS)
 	$(RM) -f ./test_thr_alarm.c
 
-test_thr_lock$(EXEEXT): thr_lock.c $(LIBRARIES)
+test_thr_lock$(EXEEXT): thr_lock.c $(LTLIBRARIES)
 	$(CP) $(srcdir)/thr_lock.c test_thr_lock.c
 	$(LINK) $(FLAGS) -DMAIN  ./test_thr_lock.c $(LDADD) $(LIBS)
 	$(RM) -f ./test_thr_lock.c
 
-test_vsnprintf$(EXEEXT): my_vsnprintf.c $(LIBRARIES)
+test_vsnprintf$(EXEEXT): my_vsnprintf.c $(LTLIBRARIES)
 	$(CP) $(srcdir)/my_vsnprintf.c test_vsnprintf.c
 	$(LINK) $(FLAGS) -DMAIN ./test_vsnprintf.c $(LDADD) $(LIBS)
 	$(RM) -f test_vsnprintf.c
 
-test_io_cache$(EXEEXT): mf_iocache.c $(LIBRARIES)
+test_io_cache$(EXEEXT): mf_iocache.c $(LTLIBRARIES)
 	$(CP) $(srcdir)/mf_iocache.c test_io_cache.c
 	$(LINK) $(FLAGS) -DMAIN ./test_io_cache.c $(LDADD) $(LIBS)
 	$(RM) -f test_io_cache.c
 
-test_dir$(EXEEXT): test_dir.c $(LIBRARIES)
+test_dir$(EXEEXT): test_dir.c $(LTLIBRARIES)
 	$(LINK) $(FLAGS) -DMAIN $(srcdir)/test_dir.c $(LDADD) $(LIBS)
 
-test_charset$(EXEEXT): test_charset.c $(LIBRARIES)
+test_charset$(EXEEXT): test_charset.c $(LTLIBRARIES)
 	$(LINK) $(FLAGS) -DMAIN $(srcdir)/test_charset.c $(LDADD) $(LIBS)
 
-testhash$(EXEEXT): testhash.c $(LIBRARIES)
+testhash$(EXEEXT): testhash.c $(LTLIBRARIES)
 	$(LINK) $(FLAGS) -DMAIN $(srcdir)/testhash.c $(LDADD) $(LIBS)
 
-test_gethwaddr$(EXEEXT): my_gethwaddr.c $(LIBRARIES)
+test_gethwaddr$(EXEEXT): my_gethwaddr.c $(LTLIBRARIES)
 	$(CP) $(srcdir)/my_gethwaddr.c ./test_gethwaddr.c
 	$(LINK) $(FLAGS) -DMAIN  ./test_gethwaddr.c $(LDADD) $(LIBS)
 	$(RM) -f ./test_gethwaddr.c
 
-test_base64$(EXEEXT): base64.c $(LIBRARIES)
+test_base64$(EXEEXT): base64.c $(LTLIBRARIES)
 	$(CP) $(srcdir)/base64.c ./test_base64.c
 	$(LINK) $(FLAGS) -DMAIN  ./test_base64.c $(LDADD) $(LIBS)
 	$(RM) -f ./test_base64.c
diff -urN mysql-orig/regex/Makefile.am mysql/regex/Makefile.am
--- mysql-orig/regex/Makefile.am	2010-01-16 15:17:39.953475662 -0100
+++ mysql/regex/Makefile.am	2010-01-16 15:17:50.275352046 -0100
@@ -16,10 +16,10 @@
 # MA 02111-1307, USA
 
 INCLUDES =		-I$(top_builddir)/include -I$(top_srcdir)/include
-noinst_LIBRARIES =	libregex.a
-LDADD=			libregex.a $(top_builddir)/strings/libmystrings.a
+noinst_LTLIBRARIES =	libregex.la
+LDADD=			libregex.la $(top_builddir)/strings/libmystrings.la
 noinst_HEADERS =	cclass.h cname.h regex2.h utils.h engine.c my_regex.h
-libregex_a_SOURCES =	regerror.c regcomp.c regexec.c regfree.c reginit.c
+libregex_la_SOURCES =	regerror.c regcomp.c regexec.c regfree.c reginit.c
 noinst_PROGRAMS =	re
 re_SOURCES =		split.c debug.c main.c
 re_LDFLAGS= @NOINST_LDFLAGS@
diff -urN mysql-orig/strings/Makefile.am mysql/strings/Makefile.am
--- mysql-orig/strings/Makefile.am	2010-01-16 15:17:39.967476070 -0100
+++ mysql/strings/Makefile.am	2010-01-16 15:17:50.275352046 -0100
@@ -16,7 +16,7 @@
 # This file is public domain and comes with NO WARRANTY of any kind
 
 INCLUDES =		-I$(top_builddir)/include -I$(top_srcdir)/include
-pkglib_LIBRARIES =	libmystrings.a
+noinst_LTLIBRARIES =	libmystrings.la
 
 # Exact one of ASSEMBLER_X
 if ASSEMBLER_x86
@@ -37,7 +37,7 @@
 endif
 endif
 
-libmystrings_a_SOURCES = $(ASRCS) $(CSRCS)
+libmystrings_la_SOURCES = $(ASRCS) $(CSRCS)
 noinst_PROGRAMS = conf_to_src
 CLEANFILES = str_test uctypedump test_decimal
 # Default charset definitions
@@ -56,28 +56,28 @@
 			t_ctype.h my_strchr.c CMakeLists.txt \
 			CHARSET_INFO.txt
 
-libmystrings_a_LIBADD=
-conf_to_src_SOURCES = conf_to_src.c xml.c ctype.c bcmp.c
-conf_to_src_LDADD=
+libmystrings_la_LIBADD=
+conf_to_src_SOURCES = conf_to_src.c
+conf_to_src_LDADD= libmystrings.la
 #force static linking of conf_to_src - essential when linking against
 #custom installation of libc
-conf_to_src_LDFLAGS= @NOINST_LDFLAGS@
+#conf_to_src_LDFLAGS= @NOINST_LDFLAGS@
 
 # This is because the dependency tracking misses @FOO@ vars in sources.
 #strtoull.o:		@CHARSET_OBJS@
 
 
-FLAGS=$(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS) @NOINST_LDFLAGS@
+FLAGS=$(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS)
 
-str_test: str_test.c $(pkglib_LIBRARIES)
-	$(LINK) $(FLAGS) -DMAIN $(INCLUDES) $(srcdir)/str_test.c $(LDADD) $(pkglib_LIBRARIES)
+str_test: str_test.c $(pkglib_LTLIBRARIES)
+	$(LINK) $(FLAGS) -DMAIN $(INCLUDES) $(srcdir)/str_test.c $(LDADD) $(pkglib_LTLIBRARIES)
 
 uctypedump: uctypedump.c
 	$(LINK) $(INCLUDES) $(srcdir)/uctypedump.c
 
-test_decimal$(EXEEXT): decimal.c $(pkglib_LIBRARIES)
+test_decimal$(EXEEXT): decimal.c $(pkglib_LTLIBRARIES)
 	$(CP) $(srcdir)/decimal.c ./test_decimal.c
-	$(LINK) $(FLAGS) -DMAIN  ./test_decimal.c $(LDADD) $(pkglib_LIBRARIES)
+	$(LINK) $(FLAGS) -DMAIN  ./test_decimal.c $(LDADD) $(pkglib_LTLIBRARIES)
 	$(RM) -f ./test_decimal.c
 
 # Don't update the files from bitkeeper
