--- mysql/libmysql_r/Makefile.am	2007-01-01 20:44:03.000000000 +0100
+++ mysql.new/libmysql_r/Makefile.am	2007-01-02 10:38:08.000000000 +0100
@@ -22,7 +22,7 @@
 
 target = libmysqlclient_r.la
 target_defs = -DDONT_USE_RAID -DMYSQL_CLIENT @LIB_EXTRA_CCFLAGS@
-LIBS = @LIBS@ @ZLIB_LIBS@ @openssl_libs@
+LIBS = @CLIENT_LIBS@ @ZLIB_LIBS@ @openssl_libs@
 
 INCLUDES =	@MT_INCLUDES@ \
 		-I$(top_builddir)/include -I$(top_srcdir)/include \
