--- mysql.orig/mysql-test/t/rpl_rotate_logs.test	2006-03-04 21:45:38.000000000 +0100
+++ mysql.fix/mysql-test/t/rpl_rotate_logs.test	2006-04-17 21:33:57.000000000 +0200
@@ -24,7 +24,7 @@
 # START SLAVE will fail because it can't read the file (mode 000)
 # (system error 13)
 --replace_result $MYSQL_TEST_DIR TESTDIR
---error 1105,1105,29
+--error 1201,1105,29
 start slave;
 system chmod 600  $MYSQLTEST_VARDIR/slave-data/master.info;
 # It will fail again because the file is empty so the slave cannot get valuable
