diff -Nru mysql-4.0.26/sql/sql_lex.cc mysql-4.0.26-logsec/sql/sql_lex.cc
--- mysql-4.0.26/sql/sql_lex.cc	2005-09-03 00:38:09.000000000 +0200
+++ mysql-4.0.26-logsec/sql/sql_lex.cc	2006-04-07 22:32:09.000000000 +0200
@@ -802,6 +802,9 @@
       while (lex->ptr != lex->end_of_query &&
 	     ((c=yyGet()) != '*' || yyPeek() != '/'))
       {
+        if (c == '\0')
+          return(ABORT_SYM);              // NULLs illegal even in comments
+
 	if (c == '\n')
 	  lex->yylineno++;
       }
