diff -Nru mysql-4.1.18/sql/sql_lex.cc mysql-4.1.18-logsec/sql/sql_lex.cc
--- mysql-4.1.18/sql/sql_lex.cc	2006-01-27 17:26:42.000000000 +0100
+++ mysql-4.1.18-logsec/sql/sql_lex.cc	2006-04-07 22:32:51.000000000 +0200
@@ -912,6 +912,9 @@
       while (lex->ptr != lex->end_of_query &&
 	     ((c=yyGet()) != '*' || yyPeek() != '/'))
       {
+        if (c == '\0')
+          return(ABORT_SYM);              // NULLs illegal even in comments
+
 	if (c == '\n')
 	  lex->yylineno++;
       }
