--- mysql-old/mysql-test/t/rpl000015.test	2006-04-18 23:11:26.000000000 +0200
+++ mysql-new/mysql-test/t/rpl000015.test	2006-04-18 23:11:44.000000000 +0200
@@ -12,7 +12,7 @@
 
 change master to master_host='127.0.0.1';
 # The following needs to be cleaned up when change master is fixed
---replace_result $MYSQL_TCP_PORT MASTER_PORT
+--replace_result $MYSQL_TCP_PORT MASTER_PORT 3306 MASTER_PORT
 --replace_column 1 # 8 # 9 # 23 # 33 #
 show slave status;
 --replace_result $MASTER_MYPORT MASTER_PORT
